### PR TITLE
I18n error paths

### LINKED ISSUE #10320: python exception from i18n module when 'msgfmt' command is not found
**Describe the bug**
When running meson it detected that msgfmt was not found, warning that translations would be skipped, but then promptly threw a python stack trace and died:

```
Program msgfmt found: NO
po/meson.build:6: WARNING: Gettext not found, all translation targets will be ignored.
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/mesonbuild/mesonmain.py", line 153, in run
    return options.run_func(options)
  File "/usr/lib/python3/dist-packages/mesonbuild/msetup.py", line 301, in run
    app.generate()
  File "/usr/lib/python3/dist-packages/mesonbuild/msetup.py", line 185, in generate
    self._generate(env)
  File "/usr/lib/python3/dist-packages/mesonbuild/msetup.py", line 229, in _generate
    intr.run()
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreter/interpreter.py", line 2700, in run
    super().run()
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/interpreterbase.py", line 149, in run
    self.evaluate_codeblock(self.ast, start=1)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/interpreterbase.py", line 174, in evaluate_codeblock
    raise e
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/interpreterbase.py", line 167, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/interpreterbase.py", line 180, in evaluate_statement
    return self.function_call(cur)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/interpreterbase.py", line 455, in function_call
    res = func(node, func_args, kwargs)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/decorators.py", line 277, in wrapper
    return f(*nargs, **wrapped_kwargs)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/decorators.py", line 574, in wrapper
    return f(*wrapped_args, **wrapped_kwargs)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreter/interpreter.py", line 2200, in func_subdir
    self.evaluate_codeblock(codeblock)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/interpreterbase.py", line 174, in evaluate_codeblock
    raise e
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/interpreterbase.py", line 167, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/interpreterbase.py", line 190, in evaluate_statement
    return self.evaluate_if(cur)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/interpreterbase.py", line 273, in evaluate_if
    self.evaluate_codeblock(i.block)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/interpreterbase.py", line 174, in evaluate_codeblock
    raise e
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/interpreterbase.py", line 167, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/interpreterbase.py", line 184, in evaluate_statement
    return self.method_call(cur)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/interpreterbase.py", line 483, in method_call
    res = obj.method_call(method_name, args, kwargs)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreter/interpreterobjects.py", line 746, in method_call
    ret = method(state, args, kwargs)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/decorators.py", line 666, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/decorators.py", line 64, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/decorators.py", line 574, in wrapper
    return f(*wrapped_args, **wrapped_kwargs)
  File "/usr/lib/python3/dist-packages/mesonbuild/modules/i18n.py", line 181, in merge_file
    '--msgfmt=' + self.tools['msgfmt'].get_path(),
TypeError: can only concatenate str (not "NoneType") to str

data/meson.build:4:7: ERROR: Unhandled python exception

    This is a Meson bug and should be reported!
```

The build statement at ``data/meson.build:4:7`` was simply this:

```
  i18n.merge_file (
    type: 'desktop',
    input: desktop + '.in',
    output: desktop,
    po_dir: po_dir,
    install: true,
    install_dir: join_paths(datadir, 'applications')
  )
```

I hit this on Debian Sid, with meson 0.62.1 when I had failed to install `gettext`

**To Reproduce**

Experimentally, if you *only* have a call to ``i18n.merge_file()`` it works fine, but if you *first* have a call to ``i18n.gettext()`` then the subsequent ``i18n.merge_file()`` throws the python exception

```
# cat meson.build 


project(
  'demo','c',
  version: '1.0',
  license: 'GPLv3+',
)

i18n = import('i18n')

i18n.gettext(meson.project_name(),
             args: ['--sort-output'],
             preset: 'glib')

i18n.merge_file (
    type: 'desktop',
    input: 'demo.desktop.in',
    output: 'demo.desktop',
    po_dir: 'po',
    )

# meson buid
The Meson build system
Version: 0.62.1
Source dir: /
Build dir: /buid
Build type: native build
Project name: demo
Project version: 1.0
C compiler for the host machine: cc (gcc 11.3.0 "cc (Debian 11.3.0-1) 11.3.0")
C linker for the host machine: cc ld.bfd 2.38
Host machine cpu family: x86_64
Host machine cpu: x86_64
Program msgfmt found: NO
meson.build:13: WARNING: Gettext not found, all translation targets will be ignored.
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/mesonbuild/mesonmain.py", line 153, in run
    return options.run_func(options)
  File "/usr/lib/python3/dist-packages/mesonbuild/msetup.py", line 301, in run
    app.generate()
  File "/usr/lib/python3/dist-packages/mesonbuild/msetup.py", line 185, in generate
    self._generate(env)
  File "/usr/lib/python3/dist-packages/mesonbuild/msetup.py", line 229, in _generate
    intr.run()
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreter/interpreter.py", line 2700, in run
    super().run()
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/interpreterbase.py", line 149, in run
    self.evaluate_codeblock(self.ast, start=1)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/interpreterbase.py", line 174, in evaluate_codeblock
    raise e
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/interpreterbase.py", line 167, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/interpreterbase.py", line 184, in evaluate_statement
    return self.method_call(cur)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/interpreterbase.py", line 483, in method_call
    res = obj.method_call(method_name, args, kwargs)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreter/interpreterobjects.py", line 746, in method_call
    ret = method(state, args, kwargs)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/decorators.py", line 666, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/decorators.py", line 64, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/usr/lib/python3/dist-packages/mesonbuild/interpreterbase/decorators.py", line 574, in wrapper
    return f(*wrapped_args, **wrapped_kwargs)
  File "/usr/lib/python3/dist-packages/mesonbuild/modules/i18n.py", line 181, in merge_file
    '--msgfmt=' + self.tools['msgfmt'].get_path(),
TypeError: can only concatenate str (not "NoneType") to str

meson.build:15:5: ERROR: Unhandled python exception

    This is a Meson bug and should be reported!

```

**Expected behavior**
It should report that 'msgfmt' is not found & exit cleanly when calling ``i18n.merge_file``.  This happens correctly today *provided* that ``i18n.gettext`` is not called first.

```
# cat meson.build 


project(
  'demo','c',
  version: '1.0',
  license: 'GPLv3+',
)

i18n = import('i18n')


i18n.merge_file (
    type: 'desktop',
    input: 'demo.desktop.in',
    output: 'demo.desktop',
    po_dir: 'po',
    )

# meson buid
The Meson build system
Version: 0.62.1
Source dir: /
Build dir: /buid
Build type: native build
Project name: demo
Project version: 1.0
C compiler for the host machine: cc (gcc 11.3.0 "cc (Debian 11.3.0-1) 11.3.0")
C linker for the host machine: cc ld.bfd 2.38
Host machine cpu family: x86_64
Host machine cpu: x86_64
Program msgfmt found: NO

meson.build:12:5: ERROR: Program 'msgfmt' not found or not executable

A full log can be found at /buid/meson-logs/meson-log.txt
```

**system parameters**
Running in a debian:sid docker container with only 'apt-get update && apt-get install meson'  run, giving a minimal env
 
# meson --version
0.62.1
# ninja --version
1.10.1
# cat /etc/deb
debconf.conf    debian_version  
# cat /etc/debian_version 
bookworm/sid
# python3 --version
Python 3.10.4

### PR DESCRIPTION
**Pull request by eli-schwartz**
- correctly check for msgfmt in merge_file() after gettext()
  Fixes #10320
- allow NLS to still work with gettext-tiny, even though maintainer targets are broken
  /cc @q66

### CODE REVIEW

**Comment by q66** [comment]
unfortunately the system is not far enough yet to set up CI from :p but i can more or less promise this will remain tested in practice

**Review by dcbaker — DISMISSED** [review]
I have one small nit, but otherwise this looks like a good change to me.

**Comment by awilfox** [comment]
We're running into this at Adélie after updating Meson to 0.62.1.  We only ship `gettext-tiny`; GNU gettext isn't even packaged.  We're working to release container images which would provide another environment you could use for CI.

(Our internal CI is definitely going to keep this patch warm, too.)

File: mesonbuild/modules/i18n.py [hunk 1]
```diff
@@ -223,13 +219,18 @@ def merge_file(self, state: 'ModuleState', args: T.List['TYPE_var'], kwargs: 'Me
         ),
     )
     def gettext(self, state: 'ModuleState', args: T.Tuple[str], kwargs: 'Gettext') -> ModuleReturnValue:
-        for tool in ['msgfmt', 'msginit', 'msgmerge', 'xgettext']:
+        for tool, strict in {'msgfmt': True, 'msginit': False, 'msgmerge': False, 'xgettext': False}.items():
```
[the remark this answers is not in the document]
Comment (eli-schwartz) [reply mesonbuild/modules/i18n.py]: Fixed.

**Comment by eli-schwartz** [comment]
@nirbheek sorry, the first commit in this series is more of a cleanup and doesn't have an associated stable regression.

And it changes the CLI interface of `meson --internal gettext`, which means that if it is backported:
- a build directory configured with 0.62.2, and built with `ninja` after downgrading to 0.62.1, will error out with invalid argparse flags
- a build directory configured with 0.62.1, and built with `ninja` after upgrading to 0.62.1, will error out because we now assume some argparse flag is passed, and don't fall back to environment variables

So we cannot backport it. Only the other two commits should be backported, as those actually do fix obscure bugs in the released 0.62.1

**Comment by eli-schwartz** [comment]
I'm pretty sure this is NOT a regression, because previously it used `$MESON_SOURCE_ROOT` which also points to the global source root, meaning this has never worked in any version of Meson.

Furthermore the target in question is a maintainer target (it only affects xgettext used for extracting message strings), not needed for build+install, so I suspect it didn't usually get run for subprojects :) which may explain why no one noticed it before or thought to do anything about it.

**Pull request merged**